ostd/src/orpc/oqueue/ringbuffer/mpmc.rs

```diff
@@ -0,0 +1,564 @@
+// SPDX-License-Identifier: MPL-2.0
+//! A MPMC OQueue implementation with support for strong and weak observation with both static and dynamic configurability.
+
+use alloc::{
+    borrow::ToOwned,
+    boxed::Box,
+    sync::{Arc, Weak},
+    vec::Vec,
+};
+use core::{
+    any::type_name,
+    cell::{Cell, UnsafeCell},
+    marker::PhantomData,
+    panic::{RefUnwindSafe, UnwindSafe},
+    ptr,
+    sync::atomic::{AtomicU64, AtomicUsize, Ordering},
+};
+
+use crossbeam_utils::CachePadded;
+
+use super::Element;
+use crate::{
+    orpc::oqueue::{
+        Blocker, Consumer, Cursor, OQueue, OQueueAttachError, Producer, StrongObserver,
+        WeakObserver,
+    },
+    sync::{Mutex, WaitQueue, Waker},
+};
+
+struct Slot<T> {
+    turn: CachePadded<AtomicUsize>,
+    value: Element<T>,
+}
+
+impl<T> Slot<T> {
+    fn new() -> Self {
+        Slot {
+            turn: CachePadded::new(AtomicUsize::new(0)),
+            value: Element::uninit(),
+        }
+    }
+
+    unsafe fn store(&self, data: T) {
+        unsafe { (*self.value.data.get()).write(data) };
+    }
+
+    unsafe fn get(&self) -> T {
+        // SAFETY: This assumes that the self.store has be previously called
+        unsafe {
+            let data = ptr::read(self.value.data.get()).assume_init();
+            // TODO(aneesh): is the following necessary?
+            // (*self.value.data.get()).assume_init_drop();
```

Comment:
I think doing this would be undefined behavior because the line above takes ownership and T may not be `Copy`.